### PR TITLE
Fixed issues with ordered endpoints

### PR DESCRIPTION
**Pull request by Alopalao**
Related to kytos PR [#221](https://github.com/kytos-ng/kytos/pull/329)

### Summary

Adapting `path` and `EVCDeploy` to ordered endpoints from kytos PR. Endpoints are chosen  depending on first/previous switch which is usually `uni_a` switch.

### Local Tests

Installing EVCs and disabling links

### End-to-End Tests

```
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /tests
plugins: timeout-2.1.0, rerunfailures-10.2, anyio-3.6.2
collected 232 items

tests/test_e2e_01_kytos_startup.py ..                                    [  0%]
tests/test_e2e_05_topology.py ..................                         [  8%]
tests/test_e2e_10_mef_eline.py ..........ss.....x.....x.RRFRRF.......... [ 24%]
...                                                                      [ 25%]
tests/test_e2e_11_mef_eline.py ......                                    [ 28%]
tests/test_e2e_12_mef_eline.py .....Xx.                                  [ 31%]
tests/test_e2e_13_mef_eline.py .....xs.s......xs.s.XXxX.xxxx..X......... [ 49%]
...                                                                      [ 50%]
tests/test_e2e_14_mef_eline.py x                                         [ 51%]
tests/test_e2e_15_mef_eline.py ..                                        [ 52%]
tests/test_e2e_20_flow_manager.py .....................                  [ 61%]
tests/test_e2e_21_flow_manager.py ...                                    [ 62%]
tests/test_e2e_22_flow_manager.py ...............                        [ 68%]
tests/test_e2e_23_flow_manager.py ..............                         [ 75%]
tests/test_e2e_30_of_lldp.py ....                                        [ 76%]
tests/test_e2e_31_of_lldp.py ...                                         [ 78%]
tests/test_e2e_32_of_lldp.py ...                                         [ 79%]
tests/test_e2e_40_sdntrace.py ...........                                [ 84%]
tests/test_e2e_41_kytos_auth.py ........                                 [ 87%]
tests/test_e2e_50_maintenance.py ........................                [ 97%]
tests/test_e2e_60_of_multi_table.py .....                                [100%]
```
Both fails are from unordered paths, they need to be ordered in e2e repository.
These failing tests are:
- `def test_140_current_path_value_given_dynamic_backup_path_and_primary_path_conditions(self)`
- `def test_145_current_path_value_given_dynamic_backup_path_and_empty_primary_conditions(self)`

Fixed on PR [#244](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/244)

### CODE REVIEW

File: models/path.py [hunk 1]
```diff
@@ -48,11 +48,13 @@ def is_valid(self, switch_a, switch_z, is_scheduled=False):
         """Check if this is a valid path."""
         if not self:
             return True
-        previous = switch_a
+        previous = {switch_a}
```
Comment (viniarck) [inline models/path.py:51]: Although initially, [I also had the impression that we could pull off accepting any order of the endpoints](https://github.com/kytos-ng/kytos/pull/329#issuecomment-1425958279) when validation a previous it can result in a non contiguous path, for instance a ring topology:

```
{
    "name": "epl_static",
    "dynamic_backup_path": false,
    "uni_a": {
        "interface_id": "00:00:00:00:00:00:00:01:1"
    },
    "uni_z": {
        "interface_id": "00:00:00:00:00:00:00:03:1"
    },
    "primary_path": [
        {
            "endpoint_a": {
                "id": "00:00:00:00:00:00:00:01:3"
            },
            "endpoint_b": {
                "id": "00:00:00:00:00:00:00:02:2"
            }
        },
        {
            "endpoint_a": {
                "id": "00:00:00:00:00:00:00:01:4"
            },
            "endpoint_b": {
                "id": "00:00:00:00:00:00:00:03:3"
            }
        }
    ]
}
```

Which would result in a broken dataplane since it would forward traffic to s2, which would end up being dropped, and exceptionally in this case the last link was also valid connecting to the last switch, so the path isn't actually contiguous:



So, the order of the endpoints actually matter to validate if the path is contigious, good thing though that the links will have deterministic endpoint with your contribution, so it's a just a matter of adjusting the algorithm here again, could you adapt it? Let's also cover this case with a unit test.

**Review by viniarck — CHANGES_REQUESTED** [review]
Great PR @Alopalao, and appreciated your help with this one. However, I've caught a corner case that's a regression, check out the comment that I made.

File: models/path.py (hunk 1 above)
Comment (viniarck) [reply models/path.py:51]: Also, since having deterministic `Link` endpoints will lose the semantics of ordering, maybe it's not worth pursuing this new approach and maintain as it is and not fix https://github.com/kytos-ng/kytos/issues/221 in the first place? As @italovalcy mentioned in the issue the impact was just mostly convenience. Once links [are loaded they'll maintain the order they we're first created](https://github.com/kytos-ng/topology/blob/master/main.py#L125). So, still having this instantiation order of links still allow mef eline to perform these static validations.

Maybe it's worth to maintain as it is, let me know what you guys think? Unless @Alopalao you can spot another solution without having to call pathfinder for static paths.

Comment (Alopalao) [reply models/path.py:51]: Fixed false positive for paths. I think the proposed changed can work as long as the path is analyzed from start/end point in order to determine whether a path is connected.

File: models/path.py [hunk 2]
```diff
@@ -50,11 +50,12 @@ def is_valid(self, switch_a, switch_z, is_scheduled=False):
             return True
         previous = {switch_a}
         for link in self:
-            current = {link.endpoint_a.switch, link.endpoint_b.switch}
-            if len(previous & current) != 1:
+            current = {link.endpoint_a.switch, link.endpoint_b.switch} \
+                      - previous
```
Comment (viniarck) [inline models/path.py:54]: @Alopalao indeed, using the `previous` here was great. But maybe for completeness you could also ensure that it's a loop free path, keeping track of all previous visited dpids? For dynamic paths with `pathfinder` when computing the shortest path we already have that guarantee with Dijkstra's. Other than that it looks great to me, could you also explore a static path that might be looped?

Comment (viniarck) [reply models/path.py:54]: Also, just noticed a trivial typo below in the `InvalidPath` `"Previus"`, if you could also patch that up, thanks.

File: models/path.py (hunk 1 above)
Comment (viniarck) [reply models/path.py:51]: Good one @Alopalao let's continue the thread on this commment here https://github.com/kytos-ng/mef_eline/pull/339#discussion_r1237395453

File: models/path.py (hunk 2 above)
Comment (viniarck) [reply models/path.py:54]: Nice. Looks good. Let's close this thread.

**Review by viniarck — APPROVED** [review]
@Alopalao, excellent solution and fix :)

I'll leave it pre-approved, don't forget to update the changelog, I think this can also land right away to be part of `2023.1`. Also could you rerun e2e but only selecting mef_eline tests? Once it's passing, let's merge these PRs.

**Comment by Alopalao** [comment]
Ran modified e2e tests PR [#244](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/244) and with latest changes
```
Ran 'hello' command on MongoDB successfully. It's ready!
+ python3 -m pytest tests/test_e2e_14_mef_eline.py tests/test_e2e_10_mef_eline.py tests/test_e2e_11_mef_eline.py tests/test_e2e_12_mef_eline.py tests/test_e2e_13_mef_eline.py tests/test_e2e_15_mef_eline.py --reruns 2 -r fEr
============================= test session starts ==============================
platform linux -- Python 3.9.2, pytest-7.2.0, pluggy-1.0.0
rootdir: /tests
plugins: timeout-2.1.0, rerunfailures-10.2, anyio-3.6.2
collected 101 items

tests/test_e2e_14_mef_eline.py x                                         [  0%]
tests/test_e2e_10_mef_eline.py ..........ss.....x.....x................  [ 40%]
tests/test_e2e_11_mef_eline.py ......                                    [ 46%]
tests/test_e2e_12_mef_eline.py .....Xx.                                  [ 54%]
tests/test_e2e_13_mef_eline.py .....xs.s......xs.s.XXxX.xxxx..X......... [ 95%]
...                                                                      [ 98%]
tests/test_e2e_15_mef_eline.py ..                                        [100%]

```
